File: 155-min-stack/min-stack.js
/*
Design a stack that supports push, pop, top,

Retrieving the minimum element in constant time.

pop, top and getMin operations will always be called on non-empty stacks.
    -O(1) 
*/
var MinStack = function() {
    this.stack = []
    this.min = []
};

/** 
 * @param {number} val
 * @return {void}
 */
MinStack.prototype.push = function(val) {
    // when pushing to stack we should update the min value
    if (this.stack.length == 0) {
        this.stack.push(val)
        this.min.push(val)
    } else {
        this.min.push(Math.min(val, this.min[this.min.length - 1]))
        this.stack.push(val)
    }
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    this.stack.pop()
    this.min.pop()
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    return this.stack[this.stack.length - 1]
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    return this.min[this.min.length - 1]
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(val)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: 155-min-stack/min-stack.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.min_vals = []

    def push(self, val: int) -> None:
        self.stack.append(val)
        if len(self.min_vals) == 0:
            min_so_far = val
        else:
           min_so_far = min(self.min_vals[-1],val)
        self.min_vals.append(min_so_far)

    def pop(self) -> None:
        self.stack.pop()
        self.min_vals.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_vals[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()